realtek: clock driver for RTL83XX (v9)

plappermaul opened the pull request:
Tested on DGS-1210-20. Maybe far from perfect but I like to get feedback. Once applied we get the following new features:

```
root@OpenWrt:/# cat /sys/kernel/debug/clk/clk_summary
                       enable  prepare  protect                                duty
   clock                count    count    count        rate   accuracy phase  cycle
---------------------------------------------------------------------------------------------
 oscillator                 1        1        0    25000000          0     0  50000
    lxb_clk                 2        2        0   200000000          0     0  50000
    mem_clk                 0        0        0   300000000          0     0  50000
    cpu_clk                 0        0        0   500000000          0     0  50000
```

```
root@OpenWrt:/# dmesg | grep MHz
[    0.000000] Frequencies: CPU:500MHz LXB:200MHz MEM:300MHz
```

**Plus**: The clock driver already has initial code to reduce the RTL838X CPU clock down to 300MHz. This can be used by some cpufreq driver in the future.

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/setup.c:
```diff
@@ -13,6 +13,7 @@
 #include <linux/init.h>
 #include <linux/clkdev.h>
 #include <linux/clk-provider.h>
+#include <linux/clk.h>
```
nit; i think you'd want to sort this above clkdev.h to keep alphabetical sorting (which the rest of this file doesn't follow anyway :p

plappermaul replied:
sorted

oliv3r replied:
i think your IDE sorts incorrectly? clk.h should be above clk-dev :p e.g.
```
#include <linux/clk-provider.h>
#include <linux/clk.h>
#include <linux/clkdev.h>
```
is actually what vim says it should be

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/setup.c:
```diff
@@ -28,6 +29,46 @@
+	struct device_node *np;
+	struct clk *clk;
+
+	*cpu = 500000000; // fallback
```
```suggestion
	*cpu = 500000000; /* fallback */
```
though I'd also add when/why the fallback happens ...

svanheule replied:
I would rather not have a fallback at all. This is code that is meant to run on SoCs that have different CPU clock rates. Either DT correctly specifies the clock rate, or it doesn't and then we can't assume anything (without also checking what platform we're on, which is even worse IMHO).

plappermaul replied:
fallback rewritten.

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/setup.c:
```diff
@@ -28,6 +29,46 @@
+	clk = clk_get(NULL, "cpu_clk");
+
+	if (IS_ERR(clk)) {
+		pr_err("no clock driver, using DT.\n");
```
Hmm, `pr_*` stuff was deprecated for ages, but you don't have a dev to use `dev_err` with ... I haven't seen yet where you call this function from; is it possible at all to pass a 'dev' a long? Also probably helps with the function below, where you pass NULL.

Granted, I haven't looked at clk stuff in years ...

plappermaul replied:
init stuff rewritten

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/setup.c:
```diff
@@ -28,6 +29,46 @@
+		pr_err("no clock driver, using DT.\n");
+		np = of_find_node_by_name(NULL, "cpus");
+		if (!np) {
+			pr_err("No 'cpus' DT node, using default.");
```
```suggestion
			pr_err("No 'cpus' DT node, using default.\n");
```
Using default what, default CPU or default CPU freq, if the later; maybe print it to inform the user?
```suggestion
			pr_err("No 'cpus' DT node, using default: '%ld'\n", cpu);
```

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/setup.c:
```diff
@@ -50,24 +91,19 @@ void __init plat_mem_setup(void)
-			pr_info("CPU frequency from device tree: %dMHz", freq / 1000000);
-		of_node_put(np);
-	}
+	plat_get_frequencies(&cpu, &lxb, &mem);
```
arg, and now I see also that this is arch specific setup code, well at still a good improvement :)

Does mips not use normal drivers/clk/rtl838x?

svanheule replied:
OpenWrt isn't on MIPS_GENERIC yet, so we can't really use this upstream. How are early clock dependencies for CPUs normally handled? I'd imagine at least the code for `fixed-clock` needs to allow for this (since I use this with the MIPS_GENERIC kernel), so maybe that can provide some hints on how this can be done without making it platform-specific.

plappermaul replied:
rewritten

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/setup.c:
```diff
@@ -50,24 +91,19 @@ void __init plat_mem_setup(void)
-	}
+	plat_get_frequencies(&cpu, &lxb, &mem);
+	if (lxb && mem)
+		pr_info("Frequencies: CPU:%dMHz LXB:%dMHz MEM:%dMHz\n",
```
```suggestion
		pr_info("Frequencies: CPU: %d MHz LXB: %d MHz MEM: %d MHz\n",
```
makes it human readable :p

plappermaul replied:
rewritten

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/setup.c:
```diff
@@ -50,24 +91,19 @@ void __init plat_mem_setup(void)
+		pr_info("Frequencies: CPU:%dMHz LXB:%dMHz MEM:%dMHz\n",
+			cpu / 1000000, lxb / 1000000, mem / 1000000);
+	else
+		pr_info("Frequencies: CPU:%dMHz\n", cpu / 1000000);
```
```suggestion
		pr_info("Frequencies: CPU: %d MHz\n", cpu / 1000000);
```

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/* Realtek RTL83XX clock driver
```
ah wait, there is a real clock driver for mips; but I suppose you it's not fair to remove the mach specific stuff to be replaced with this as part of this patch series?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+
+#include <linux/clk-provider.h>
+#include <linux/clkdev.h>
+#include <linux/clk.h>
```
Heh, my editor sorts these in the opposite order :p

plappermaul replied:
sorted

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	PLL_INFO(CLK_LXB, "lxb_clk", "ref_clk")
+};
+
+#define PLL_COUNT ARRAY_SIZE(pll_info)
```
since your using this define later, why 'hide' it here; keep it at the top?

plappermaul replied:
reordered

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	struct rtl_pll *pll = rtl_hw_to_pll(hw);
+	unsigned int pll_base = (rtl_soc_family == RTL8380_FAMILY_ID) ? rtl838x_pll_base[pll->id] : rtl839x_pll_base[pll->id];
+
+	unsigned int ctrl0 = sw_r32(pll_base);
```
I know we use a lot of `sw_r32` in our rtl-codebase; but isn't there a nicer and cleaner kernel interface for this?

btw, why are you not getting the pll_base from the devicetree?

svanheule replied:
`readl()` and `ioread32()` should do the same thing. The PLL registers are in the switchcore area, although they do form a single cluster https://svanheule.net/realtek/maple/feature/pll_bias

plappermaul replied:
Better stay with sw_XX() stuff.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	unsigned int pll_base = (rtl_soc_family == RTL8380_FAMILY_ID) ? rtl838x_pll_base[pll->id] : rtl839x_pll_base[pll->id];
+
+	unsigned int ctrl0 = sw_r32(pll_base);
+	unsigned int ctrl1 = sw_r32(pll_base + 4);
```
wouldn't be (both) be `pll_base + RTL83xx_PLL_CTRL1` as is more comon instead of hardcoded values?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	unsigned int ctrl0 = sw_r32(pll_base);
+	unsigned int ctrl1 = sw_r32(pll_base + 4);
+	
+	unsigned int cmu_sel_prediv = ctrl0 & 0x3;
```
include `bitops.h` and (at the top) define something like

`RTL83XX_PLL_CTRL1_CMU_SEL_PREDIV      GENMASK(1, 0)`

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	unsigned int ctrl1 = sw_r32(pll_base + 4);
+	
+	unsigned int cmu_sel_prediv = ctrl0 & 0x3;
+	unsigned int cmu_sel_div4 = ctrl0 >> 2 & 0x1;
```
you probably want to have a macro that gets this for you from ctrl; something like

```
#define RTL83XX_CTRL1_CMU_SEL_DIV4_MASK  GENMASK(3, 1)
#define RTL83XX_CTRL1_CMU_SEL_DIV4_SHIFT 2
#define RTL83XX_CTRL1_CMU_SEL_DIV4[_GET](reg) \
             (((reg) & RTL83XX_CTRL1_CMU_SEL_DIV4_MASK) >> RTL83XX_CTRL1_CMU_SEL_DIV4_SHIFT)
```
While typing out these defines is a bit of work; it makes things a lot more readable (and reusable). Especially doing your mask before your shift; means you'll genmask on EXACTLY the bits you want/need which should map the datasheet/reg description.

Also, the `_SHIFT` and `_MASK`  can nicely be used on the setters as well :)

svanheule replied:
@oliv3r That `GENMASK()` for `SEL_DIV4` doesn't look right (size 1, LSB=2) https://svanheule.net/realtek/maple/register/pll_cpu_ctrl0

With a shifted (compile time constant) mask, the following is also possible:
```
#define RTL838X_PLL_CTRL0_CMU_SEL_DIV4  BIT(2)

unsigned int cmu_sel_div4 = FIELD_GET(RTL838X_PLL_CTRL0_CMU_SEL_DIV4, ctrl0);
```

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	if (rtl_soc_family == RTL8380_FAMILY_ID) {
+		cmu_divn2_selb = ctrl1 >> 26 & 1;
+		cmu_divn3_sel = ctrl1 >> 27 & 3;
+	} else {
```
what else? what family do we have if its' NOT the RTL8380? better make it explicit with a switch case where the default/unknown case is an error.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+
+	return (((parent_rate / 1000) * (cmu_ncode_in + 4)) /
+		(((cmu_divn2_selb == 0) ? (cmu_divn2 + 4) : rtl83xx_pll_divn3[cmu_divn3_sel]) * rtl83xx_pll_xdiv[pll->id])) *
+		(1 << cmu_sel_prediv) * ((cmu_sel_div4 == 0) ? 1 : 4) * 1000;
```
oh wow, that's quite complex :) maybe split it up with a view intermediate variables (they get compiled away anyway) to keep this readble? and watch your magic values, remember you can use `BIT(cmu_sel_prediv)` if we're talking about that specifically, but a lot of magic values here :(

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+
+	unsigned int cmu_ncode_in;
+	unsigned int cmu_divn3_sel;
+	unsigned int cmu_big_kvco;
```
I suppose these have different meaning and don't need to be initialized?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	unsigned int cmu_ldo_en = 0x1;
+	unsigned int cmu_vc_dly = 0x0;
+	unsigned int cmu_en_ckoobs = 0x0;
+	unsigned int test_en = 0x0;
```
are these consts? do you modify them or only use them? if you only use them, then probably use a `#define` instead. otherwise.

You have more variables then actual code in this function :p

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	unsigned int cmu_en_ckoobs = 0x0;
+	unsigned int test_en = 0x0;
+
+	if (rate == 500000000 && parent_rate == 25000000 && pll->id == CLK_CPU) {
```
```suggestion
	if ((rate == 500000000) && (parent_rate == 25000000) && (pll->id == CLK_CPU)) {
```
don't forget your parenthesis, nobody needs to memorize operator precedence :) it'll save some bug at some point by doing this consistently :)

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+		return 0;
+	}
+
+	ctrl0 = (cmu_fcode_in<<20)|(cmu_divn2<<12)|(cmu_ncode_in<<4)|
```
checkpatch will complain about missing spaces around operators here ...

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	if (pll->id == CLK_CPU) 
+		return (rate < 450000000) ? 300000000 : 500000000;
+	else if (pll->id == CLK_MEM) 
+		return (rate < 246875000) ? 193750000 : 300000000;
```
these are very particular numbers, just curious where they are from? If they where 25Mhz and 20Mhz I wouldn't have batted an eye ...

plappermaul replied:
246875000 is half between 193750000 and 300000000 and so we are rounding up/down. Nevertheless I do not know if we will ever be able to change memory frequency to 193MHz. Next rewrite will take care about that.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+
+static int rtl838x_pll_enable(struct clk_hw *hw)
+{
+	return 0;
```
probably nice to add a comment that this is currently unknown or not possible? e.g. why do we only have stubs.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	.enable = rtl838x_pll_enable,
+	.disable = rtl838x_pll_disable,
+	.is_enabled = rtl838x_pll_is_enabled
+};
```
if you leave these undefined/unset; doesn't the clkframe work have some defaults/skip them? That also makes your above stubs not needed I think ...

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+		cmu_divn3_sel = 0x2;
+		cmu_big_kvco = 0x0;
+	} else {
+		pr_err("not allowed set clock %d to %luMHz\n", pll->id, rate / 1000000);
```
since you have your `clk_hw` here, you can derive your `dev` and can (and should) use your `dev_err` functions instead

plappermaul commented:
memo to myself: Reorder registration to match https://elixir.bootlin.com/linux/latest/source/drivers/clk/mediatek/clk-mt8192.c

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	clk_id = clkspec->args[0];
+	pll = rtl_pll_find_desc(data, clk_id);
+	if (IS_ERR(pll)) {
+		pr_info("Invalid PLL clock ID %d specified\n", clk_id);
```
there should be a dev to be found in `clkspec`

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+	if (ret)
+		goto err_free_pll;
+
+	clk_hw_register_clkdev(&pll->hw, pll_init->name, NULL);
```
do we have a managed `devm_clk_hw_register_clkdev` by any chance?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+		if (rtl_soc_family == RTL8390_FAMILY_ID)
+			init.ops = &rtl839x_pll_ops;
+		else
+			init.ops = &rtl838x_pll_ops;
```
as before, make it explicit and error out in the other case.

I wonder if we can combine the RTL930x in this driver as well; shouldn't be that unreasonable. I think the RTL931x might be tricky, as it's a different mips core; but even that might feature many similarities ...

plappermaul replied:
rewritten

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+		data->plls[idx] = rtl_pll_hw_register(&init);
+		if (IS_ERR(data->plls[idx])) {
+			ret = PTR_ERR(data->plls[idx]);
+			pr_err("Couldn't register PLL hw '%s'\n", init.name);
```
you may have to add a dev to your `data` struct ...

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,461 @@
+{
+	pr_info("RTL839X clock init");
+	rtl_clk_init_dt(np, RTL8390_FAMILY_ID);	
+}
```
I suppose this is the reason we still have the 'mach' specific functions, can't we do this via devicetree probes?

plappermaul replied:
rewritten

oliv3r commented on target/linux/realtek/files-5.10/include/dt-bindings/clock/rtl83xx-clk.h:
```diff
@@ -0,0 +1,14 @@
+ * RTL83XX clock indices
+ */
+#ifndef __DT_BINDINGS_CLOCK_RTL83XX_H
+#define __DT_BINDINGS_CLOCK_RTL83XX_H
```
```suggestion
#define __DT_BINDINGS_CLOCK_RTL83XX_H __FILE__
```
I've learned this trick a while ago and quite like it; it means it IS define, has ACTUAL content and nothing too random :)

plappermaul replied:
Linux kernel dows not use __FILE__ anywhere in the DT.

robimarko commented:
To me the clock driver looks self-contained, so why not make it a platform driver?
Then you can have a device to utilize instead of working around it.

Also, I would prefer to get rid of mach manual setup,

plappermaul commented:
Regarding the platform driver we might have a chicken-egg-problem. We want to read CPU frequencies in setup.c to define mips_hpt_frequency. At this point not platform driver has been loaded yet. Earliest possibly driver init call starts later. 

```
<<< NEEDED HERE >>>
[    0.000000] Frequencies: CPU:500MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041786 ns
[    0.000020] sched_clock: 32 bits at 250MHz, resolution 4ns, wraps every 8589934590ns
[    0.008796] Calibrating delay loop... 498.89 BogoMIPS (lpj=2494464)
[    0.075681] pid_max: default: 32768 minimum: 301
[    0.081154] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.089196] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.104132] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build

<<< EARLIEST POSSIBLE PLATFORM DRIVER LOAD >>>
[    0.112628] EARLY_INITCALL rtl83xx clock 
[    0.122426] PURE_INITCALL rtl83xx clock
[    0.127861] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.138785] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.146459] pinctrl core: initialized pinctrl subsystem
[    0.152720] CORE_INITCALL rtl83xx clock
[    0.158666] NET: Registered protocol family 16
[    0.164300] POSTCORE_INITCALL rtl83xx clock
[    0.176521] ARCH_INITCALL rtl83xx clock
[    0.338857] clocksource: Switched to clocksource MIPS
```

For comparison look at corresponding mainline ath79 setup routines. It works with CLK_OF_DECLARE() and runs totally outside the clk drivers.

https://elixir.bootlin.com/linux/latest/source/arch/mips/ath79/setup.c#L263
https://elixir.bootlin.com/linux/latest/source/arch/mips/ath79/clock.c

Any ideas?

svanheule commented:
What you need to compare against, is a `MIPS_GENERIC` based platform. The `MIPS_GENERIC` setup code will use the DT info to fetch the clock, and to initialise `mips_hpt_frequency` in a generic way. See https://elixir.bootlin.com/linux/latest/source/arch/mips/generic/init.c#L151

See for example [ingenic/jz4770.dtsi](https://elixir.bootlin.com/linux/latest/source/arch/mips/boot/dts/ingenic/jz4770.dtsi), which has a a phandle referring to the [`ingenic,jz4770-cgu`](https://elixir.bootlin.com/linux/latest/source/drivers/clk/ingenic/jz4770-cgu.c) clk driver as CPU clock.

plappermaul commented:
The current PR works exactly this way. So if nobody has a good idea I will stay away from the platform driver.

svanheule commented:
Right, `OF_CLK_DECLARE()` is the right thing to do, and is not a platform driver (just like on MIPS ingenic). Perhaps I need more coffee...

Still, we should be working towards compatibility with `MIPS_GENERIC` in my opinion, which was mostly the point I and @robimarko (IIUC) were trying to make. So if you want to show the platform clock frequencies like the SDK does, you should rather be doing this from the clk driver itself, and keep `setup.c` clean of extras. I don't think this is necessary to do this all though, because with your patches this information can be accessed from debugfs.

plappermaul commented:
I got lost in platform driver and MIPS clock initialization routines. But now I think the framework should be fine.

Find attached a partly rewirtten PR. Please comment on platform and driver initialization routines.

bkobl commented:
This is some seriously cool stuff! I am looking forward to overclocked switches with water-cooling, we shall also need support for LED effects from the port leds for the modder-scene ;-)

No, seriously, this has great potential for power saving, which is becoming ever more important these days.

I would not bother too much about MIPS_GENERIC. This ship has sailed for the RTL9300 without a MIPS timer interrupt and the way the RTL8390 and RTL9300 handle the VPEs.

You write RTL83xx clock driver, is there support for the RTL839x? My understanding is that there was somewhat an evolution from the RTL838x regarding the clocks towards the RTL839x. And the RTL93xx code looks quite different again.

For development purposes it would be interesting to be able to change the LX clock. How are peripherials going to be informed that there is a clock change, will this be possible at run-time?

From my side, the code is good to be merged, it seems to work so far, and I am sure it will spark lots of ideas.

plappermaul commented:
Off we go. PR v3 includes an overhaul of the clock driver and adds full RTL839x support.

Feel free to test & comment.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/Kconfig:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: GPL-2.0-only
+config COMMON_CLK_RTL83XX
+        bool "Clock driver for Realtek RTL83XX"
+	depends on MIPS
```
This should depend on the RTL83XX architecture here. The driver doesn't apply to MIPS in general.

plappermaul replied:
fixed in upcoming version

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl839x-sram.S:
```diff
@@ -0,0 +1,94 @@
+
+#include <dt-bindings/clock/rtl83xx-clk.h>
+
+#define RTL838X_SW_BASE		0xbb000000
```
`838X_BASE`, but the file is 839x-related.

plappermaul replied:
fixed in upcoming version

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,506 @@
+#include <dt-bindings/clock/rtl83xx-clk.h>
+#include <asm/mach-rtl838x/mach-rtl83xx.h>
+
+extern struct rtl83xx_soc_info soc_info;
```
We tried containing this to the ethernet driver [at one point](https://patchwork.ozlabs.org/project/openwrt/list/?series=276423&state=*), so I would rather not see it spread through the drivers again.

plappermaul replied:
fixed in upcoming version

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,506 @@
+
+static const struct rtl_reg_set rtl838x_cpu_reg_set[] = {
+	REG_SET(300000000, 0x045c8, 0x1414530e, 0x2d20cfb8),
+	REG_SET(500000000, 0x04748, 0x0c14530e, 0x2da0cfb8)
```
Are these the pre-combined values of all those little register fields you had in the previous version?

plappermaul replied:
yep. These are the hardware magic values. using constants is better than some ugly code combination.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,506 @@
+	unsigned int ctrl1 = sw_r32(rtl_regs[rtl_ccu->soc][REG_CTRL1][clk->idx]);
+
+	unsigned int div1, div2 = rtl_xdiv[clk->idx];
+	unsigned int cmu_sel_prediv = 1 << (ctrl0 & 0x3);
```
I would rather see something like `1 << FIELD_GET(RTL838X_CMU_SEL_PREDIV, ctrl0)` here.

plappermaul replied:
maybe we can start the beauty contest later ...

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,506 @@
+	div1 = (cmu_divn2_selb == 0) ? cmu_divn2 : cmu_divn3_sel;
+
+	return (((parent_rate / 16) * cmu_ncode_in) / (div1 * div2))
+		* cmu_sel_prediv * cmu_sel_div4 * 16;
```
Strange naming here. Divide by dividers and multiply by dividers?

plappermaul replied:
I know, But this is register field naming and so I will keep the var names.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,506 @@
+static int rtl_ccu_create(struct device_node *np)
+{
+	if (soc_info.family != RTL8380_FAMILY_ID && soc_info.family != RTL8390_FAMILY_ID)
+		return -ENXIO;
```
If you don't use `soc_info`, this can be dropped.

plappermaul replied:
fixed in upcoming version

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,506 @@
+	case RTL8390_FAMILY_ID:
+		rtl_ccu->soc = SOC_RTL839X;
+		break;
+	}
```
This, and the above check, should be replaced with device tree compatibles (like in you previous version).
```C
if (of_device_is_compatible(np, "realtek,rtl838x-clock"))
	rtl_ccu->soc = SOC_RTL838X;
else if (of_device_is_compatible(np, "realtek,rtl839x-clock"))
	rtl_ccu->soc = SOC_RTL839X;
else
	return -ENXIO;
```

The clock peripherals for these two SoC families are different, so they should definitely have different compatibles.

plappermaul replied:
fixed in upcoming version

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,506 @@
+ * devices (e.g. UART). Therefore we cannot use the platform driver framework.
+ */
+
+CLK_OF_DECLARE(rtl83xx, "realtek,rtl83xx-clock", rtl_clk_init);
```
```c
CLK_OF_DECLARE(rtl838x, "realtek,rtl838x-clock", rtl_clk_init);
CLK_OF_DECLARE(rtl839x, "realtek,rtl839x-clock", rtl_clk_init);
```
The rtl838x/rtl839x names are likely to be refused upstream though. You could use rtl8380/rtl8390 to classify these peripherals, which is what I used for the GPIO and watchdog drivers. Alternatively, you could also use Realtek's codenames "maple" and "cypress", although those make it harder to identify which drivers certain hardware would require.

plappermaul replied:
fixed in upcoming version

svanheule reviewed (changes requested):
If you get rid of the `soc_info` dependency, this driver will run without much further modification (platform specific headers) on a 5.19 kernel. Since there's no loss of functionality with the proposed changes, I really see no point in choosing not to stay compatible with the `MIPS_GENERIC` based upstream platform.

```
# uname -a && cat /sys/kernel/debug/clk/clk_summary
Linux rtl83xx 5.19.0-02010-g93c94250e0cc-dirty #182 SMP Tue Aug 2 22:05:25 CEST 2022 mips GNU/Linux
                                 enable  prepare  protect                                duty  hardware
   clock                          count    count    count        rate   accuracy phase  cycle    enable
-------------------------------------------------------------------------------------------------------
 oscillator                           1        1        0    25000000          0     0  50000         Y
    lxb_clk                           2        2        0   200000000          0     0  50000         Y
    mem_clk                           0        0        0   300000000          0     0  50000         Y
    cpu_clk                           0        0        0   500000000          0     0  50000         Y
```

bkobl commented:
> If you get rid of the `soc_info` dependency, this driver will run without much further modification (platform specific headers) on a 5.19 kernel. Since there's no loss of functionality with the proposed changes, I really see no point in choosing not to stay compatible with the `MIPS_GENERIC` based upstream platform.

Dear Sander,
the upstream platform is the upstream platform. This is not what we use on OpenWRT and it will not fly anyway for the RTL9300 and RTL839x. Asking someone to remove functionality to play your political games is absolutely not OK. At this point we do not even have 5.15 compatibility and we will likely not get there unless we get the functionality of e.g. the XGS PR in that is hanging there since some time. Your role here is to review the PR, not to impose your opinions. You do due diligence and then it gets merged.
Birger

robimarko commented:
@bkobl You got it a bit backwards, since @svanheule is one of the maintainers, his opinion pretty much equates to the project's opinion when it comes to Realtek target.
You cant just chuck in a 30+ commit PR and then expect anybody sane to rubber stamp that.

plappermaul commented:
Aside from the discussion above I updated to PR v4

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,651 @@
+	if (!sram_base) {
+		dev_warn(&rtcl_ccu->pdev->dev, "unable to allocate SRAM%s", msge);
+		goto err_put_device;
+	}
```
If I got this right, reclocking from SRAM is just disabled if the required code is absent (makes sense). If reclocking is provided, but the SRAM node is missing, you fail driver probing however. AFAICT, SRAM doesn't appear to be used for reclocking on RTL838x, so are you making driver probing on RTL838x depend on something that isn't required? Not that I'm against adding the SRAM node to the RTL838x DTSI; that would still serve as good documentation.

plappermaul replied:
838x will need SRAM in the forthcoming PR version too.

svanheule commented on target/linux/realtek/dts-5.10/rtl839x.dtsi:
```diff
@@ -87,6 +97,13 @@
 		#size-cells = <1>;
 		ranges = <0x0 0x18000000 0x10000>;
 
+		ccu: clock-controller {
```
This device's memory is not in the range covered by `/soc`. Since we don't have a real node for that space yet, it should under the root node for the time being. Other architectures seem to sometimes use a `/clocks` node, where this could be put together with the reference oscillator.

plappermaul replied:
fixed in upcoming version

svanheule reviewed (commented):
checkpatch.pl still reports some whitespace issues. The warning about long lines, and references to symbols from the ASM files can be ignored.

plappermaul commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl839x-sram.S:
```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Realtek RTL838X SRAM clock setters
```
fixed in upcoming version

plappermaul commented on target/linux/realtek/files-5.10/arch/mips/include/asm/mach-rtl838x/mach-rtl83xx.h:
```diff
@@ -202,6 +202,14 @@
+#define RTL839X_SRAM_BASE	0x9f000000
+#define RTL838X_SRAM_SIZE	(64*1024)
+#define RTL839x_SRAM_SIZE	(96*1024)
+
```
fixed in upcoming version

plappermaul commented on target/linux/realtek/rtl839x/config-5.10:
```diff
@@ -12,7 +12,7 @@ CONFIG_CLKDEV_LOOKUP=y
 CONFIG_CLKSRC_MMIO=y
 CONFIG_CLONE_BACKWARDS=y
 CONFIG_COMMON_CLK=y
-CONFIG_COMMON_CLK_BOSTON=y
```
fixed in upcoming version

plappermaul commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,651 @@
+	void *dram_start;
+	int dram_size;
+	char msge[80] = ", rate setting disabled.\n";
+	char msgi[80] = "rate setting enabled\n";
```
fixed in upcoming version

plappermaul commented:
PR v5 is ready for review. Changes since last version:

- RTL838x assembler setters (can change memory clock)
- allows overclocking
- SRAM anti-overwrite memory check
- DDR3 300 MHz minimum check
- fully self contained (uses own header file)

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,724 @@
+	int ret;
+
+	if (!rtcl_ccu) {
+		dev_err(&pdev->dev,"early initialization not run");
```
You can use dev_err_probe() here.

plappermaul replied:
cannot se how to use dev_err_probe() as it requires some error value.

svanheule replied:
```
if (!rtcl_ccu)
    return dev_err_probe(&pdev->dev, -ENXIO, "early initialization not run");
```

plappermaul replied:
applied, thanks

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -0,0 +1,66 @@
+#define RTL838X_PLL_MEM_CTRL0			(0x0FDC)
+#define RTL839X_PLL_MEM_CTRL0			(0x0048)
+#define RTL838X_PLL_MEM_CTRL1			(0x0FE0)
+#define RTL839X_PLL_MEM_CTRL1			(0x004C)
```
Now that these are in a separate file, it would make more sense to me to group these by family, instead of by type (GLB_CTRL, CPU_CTRL0, ...)

plappermaul replied:
fixed in next update

svanheule reviewed (commented):
Good progress! Can you let us know where you're done adding features? :upside_down_face:

plappermaul commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl838x-sram.S:
```diff
@@ -0,0 +1,127 @@
+	sync
+	sw	rTMP, 0(rGLB)
+	/* wait for ready */
+1:
```
fixed in next version

plappermaul commented:
PR v6 updates:

- allow speed ranges in DT
- fix instability due to unfinished cach flushes
- Finer speed steps for CPU on RTL839x (25MHz)

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+ *   clock-names = "ref_clk";
+ *   cpu-range = <MIN MAX>;
+ *   mem-range = <MIN MAX>;
+ *   lxb-range = <MIN MAX>;
```
Unless I missed something, these are non-standard properties. Why can't these limits just go in the driver? Do you want to be able to limit the clock frequencies per board?

plappermaul replied:
Absolutely right. Simple example for possible differences:

Zyxel GS1900-48: Board: RTL839x CPU:750MHz LXB:200MHz MEM:400MHz
DGS 1210-52: Board: RTL839x CPU:700MHz LXB:200MHz MEM:400MHz

svanheule replied:
That made me realise the GS1900-48 is already running faster than the 700MHz listed on the RTL8393's product page. Still, these frequencies are set up by the bootloader. For some reason, be it info from Realtek or their own testing, these boosted clock frequencies are considered stable. Would it make sense to use these boot values as an upper limit in the driver? The lower limit could still be hardcoded, as I don't think those would be determined by the boards.

plappermaul replied:
I will go with the following naming convention in the final version. max values will be possible but not set.

```
        ccu: clock-controller {
                compatible = "realtek,rtl8390-clock";
                #clock-cells = <1>;
                clocks = <&osc>;
                clock-names = "ref_clk";
                min-cpu-frequency = <425000000>;
                min-mem-frequency = <325000000>;
                min-lxb-frequency = <100000000>;
        };
```

svanheule replied:
I'm still not convinced these min/max values belong in the DT description. Debugfs entries to configure the allowed range would make sense to me, which would be initialised by the driver to safe values (hard coded or derived from boot state).

oliv3r replied:
"normally" the hardware has internal defaults. e.g. the chip refuses to operate outside of X specs, because those are design limitations. Those do not belong int he dt, but in the driver itself.

Now, we have certain boards that can (but should they? maybe leave it as stock and loose the OC ability ...) be over/underclocked outside of spec.

For the MVP/initial series, leave that complexity out. Better have all boards behave the same, without the loss of the boost, for simpler and easier code.

We can then later argue about what the best approach is to boost' frequencies on boards that support it. But this is 'basic overclocking' like stuff. Is it worth it? Probably not. Should we facilitate it, maybe. How? That's the tricky one. If you look at common desktop CPU's, this is handled in completely different ways. But lets keep that out of the discussion for now? For upper bounds, you can always make it so, unless explicitly set higher by the user, use the internal 'max' as max?

plappermaul replied:
already fixed

oliv3r commented on target/linux/realtek/files-5.10/include/dt-bindings/clock/rtl83xx-clk.h:
```diff
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * Copyright (C) 2022 Markus Stockhausen
```
I think upstream wants an e-mail address here generally

```suggestion
  * Copyright (C) 2022 Markus Stockhausen <markus.stockhausen@gmx.de>
```

oliv3r commented on target/linux/realtek/files-5.10/include/dt-bindings/clock/rtl83xx-clk.h:
```diff
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
```
While it pains me to say; and I much much more prefer this style as well; I think upstream prefers
```suggestion
// SPDX-License-Identifier: GPL-2.0-only
```
Also, its fine to mark it as `GPL-2.0-or-later`. YOUR choice as author, I personally prefer to allow an upgrade (if needed,r atter then blocking it potentially forever)

svanheule replied:
The (docs)[https://www.kernel.org/doc/html/v5.19/process/license-rules.html] say Markus used the correct style for C headers.

oliv3r replied:
Ah they made it even more confusing, why I still don't get. So `/*` for headers, `//` for C files. I see the reason for the change (good, I'm happy :p) what I don't get, why not make it consistent and just use `/*` for both. There's no _real_ C++ style comments in the kernel base, except for these SPDX identifiers, and then they're not consistent. Someone must have felt really strongly about wanting C++ headers and probably hoped this was 'a way in'. Especially considering the reason for the rollback on headers, might as well have done them both for consistency sake.

Anyway, sorry for the noise :)

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/Kconfig:
```diff
@@ -0,0 +1,7 @@
+	depends on RTL83XX
+	select SRAM
+	help
+	  This driver supports Realtek RTL83XX basic clocks.
```
checkpatch usually complains (or used to) about very short help entries. I know; adding words to please the linter is usually silly.

```suggestion
	  This driver adds support for the Realtek RTL83xx series basic clocks. This includes chips in the
	  RTL838x series, such as RTL8380, RTL8381, RTL832, as well as chips from the RTL839x series, such as RTL8390, RT8391, RTL8392, RTL8393 and RTL8396.
	  
	  When configured as a module, the module will be called clk-rtl83xx.
```

this enumerations are really useful for people going through the 'help' and trying to figure out what to enable :)

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/Makefile:
```diff
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: GPL-2.0-only
+obj-$(CONFIG_COMMON_CLK_RTL83XX) += clk-rtl83xx.o clk-rtl838x-sram.o clk-rtl839x-sram.o
```
is it worth-wile considering loading this blob via /lib/firmware instead? Now you are adding unused code (blobs) in your binary for no reason, wasting some space in the binary.

plappermaul replied:
We will waste space and add a TODO

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl838x-sram.S:
```diff
@@ -0,0 +1,139 @@
+ * Realtek RTL838X SRAM clock setters
+ * Copyright (C) 2022 Markus Stockhausen
+ *
+ * relocatable code that will run from SRAM
```
i'm a sucker for understanding the purpose in these headers. 'Why do we need this? (besides the what)

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl838x-sram.S:
```diff
@@ -0,0 +1,139 @@
+ */
+
+#include <dt-bindings/clock/rtl83xx-clk.h>
+#include "clk-rtl83xx.h"
```
``suggestion

#include "clk-rtl83xx.h"
```

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -0,0 +1,87 @@
+ */
+
+/*
+ * Several stuff
```
stuff of what lol :)

plappermaul replied:
of course the IMPORTANT stuff :P

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -0,0 +1,87 @@
+ */
+
+#define RTL_SRAM_MARKER				(0x5EAF00D5)
+#define RTL_SRAM_BASE				(0x9F000000)
```
is this the same on all realtek's? even so, the file is called RTL83xxx, and you only have RTL83[89]x stuff here, so
```suggestion
 #define RTL83XX_SRAM_BASE				(0x9F000000) 
```
indentation is a bit off, but that might be github's rendering .... (spaces are also ok for _indentation_

svanheule replied:
The line marker for patches can push a tab character into the next column.

oliv3r replied:
I don't this is it though; I think github renders 1 tab as 4 spaces generally. There's been quite a few bugs and issues on that topic.

svanheule replied:
You're right, sorry for the noise. That's why I tend to only spaces for indentation on my own code. No hassle with tab widths.

oliv3r replied:
hah, I think tabs are far superior for _indentation_; it holds meaning. spaces are perfect for indentation however. using both (tabs for indentation, spaces for alignment0 works great :)

Don't worry though; spaces for indentation AND alignment have plenty of other issues :D but lets not get ourselves distracted on that :p

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -0,0 +1,87 @@
+#define RTL839X_PLL_MEM_CTRL0			(0x0048)
+#define RTL839X_PLL_MEM_CTRL1			(0x004C)
+
+#define RTL_PLL_CTRL0_CMU_SEL_PREDIV(v)		(((v) >> 0) & 0x3)
```
any reason you are not adding defines for the 'offset' and masks?

```suggestion
#define RTL83XX_PLL_CTRL0_CMU_SEL_PREDIV_SHIFT  0
#define RTL83XX_PLL_CTRL0_CMU_SEL_PREDIV_MASK  GENMASK(1, 0)
#define RTL83XX_PLL_CTRL0_CMU_SEL_PREDIV(v) \
             (((v) & RTL83XX_PLL_CTRL0_CMU_SEL_PREDIV_MASK) >> RTL83XX_PLL_CTRL0_CMU_SEL_PREDIV_SHIFT)
```
yes it's a bit more typing, but it reads a lot better.
Also, swap your shift/mask direction, with GENMASK(endbit, startbit) having the advantage that it will match whatever the datasheet says baout which bits are being used. e.g. for the next one:

plappermaul replied:
As we include this header into the assembler too we should keep it simple.

oliv3r replied:
I don't doubt openwrt will be fine with that; but upstream probably prefes more clean more readable code in the end?

If you insist however, we can get it in first; then rework it and fix it later ;)

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -0,0 +1,87 @@
+#define RTL839X_PLL_MEM_CTRL1			(0x004C)
+
+#define RTL_PLL_CTRL0_CMU_SEL_PREDIV(v)		(((v) >> 0) & 0x3)
+#define RTL_PLL_CTRL0_CMU_SEL_DIV4(v)		(((v) >> 2) & 0x1)
```
```suggestion
#define RTL83XX_PLL_CTRL0_CMU_SEL_DIV4_SHIFT 2
#define RTL83xXXPLL_CTRL0_CMU_SEL_DIV4_MASK BIT(2) // No idea if GENMASK(2, 2) works :p
#define RTL83XX_PLL_CTRL0_CMU_SEL_DIV4(v) \
            (((v) & RTL83xXXPLL_CTRL0_CMU_SEL_DIV4_MASK) >> RTL83XX_PLL_CTRL0_CMU_SEL_DIV4_SHIFT) 
```

plappermaul replied:
again simplicity for assembler include.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -0,0 +1,87 @@
+
+#define RTL_PLL_CTRL0_CMU_SEL_PREDIV(v)		(((v) >> 0) & 0x3)
+#define RTL_PLL_CTRL0_CMU_SEL_DIV4(v)		(((v) >> 2) & 0x1)
+#define RTL_PLL_CTRL0_CMU_NCODE_IN(v)		(((v) >> 4) & 0xFF)
```
(yes, GENMASK(7, 0) is okay as is GENMASK(32, 0); it makes your intent very clear without magic values :)

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -0,0 +1,87 @@
+#define RTL838X_PLL_CTRL1_CMU_DIVN2_SELB(v)	(((v) >> 26) & 0x1)
+#define RTL838X_PLL_CTRL1_CMU_DIVN3_SEL(v)	(((v) >> 27) & 0x3)
+
+#define RTL839X_GLB_CTRL_CPU_CLKSEL_MASK	(1 << 11)
```
BIT(11) :p

plappermaul replied:
That will not work :P because we include the file into the assembler sources.

oliv3r replied:
WHy not? it's just a pre-processor macro; do we not run the pre-processor before assembly?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -0,0 +1,87 @@
+#define RTL_MC_DMCR				(0x101c)
+#define RTL_MC_DACCR				(0x1500)
+#define RTL_MC_DCDR				(0x1060)
+#define RTL_UART_BASE				(0x2000)
```
this is a weird one to have here (as is SOC base). This should really be retrieved from the DTS, but i'll probably see that when I'll look at your C code which I'll do next ;)

plappermaul replied:
will leave that here as it does not hurt.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+ * accessed. To make use of the driver add the following devices and configurations at the
+ * appropriate locations to the DT.
+ *
+ * #include <dt-bindings/clock/rtl83xx-clk.h>
```
these examples probably go into `Documentation/devicetreebindings/clk/realtek-rtl83xx` no? they don't _really_ belong here. But it might be good-enough for openwrt however, and we can 'fix' this later when upstreaming

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+#include <linux/platform_device.h>
+#include <linux/slab.h>
+
+#include <dt-bindings/clock/rtl83xx-clk.h>
```
i think this header should go before the linux, after the asm headers no?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+#include <dt-bindings/clock/rtl83xx-clk.h>
+#include "clk-rtl83xx.h"
+
+#define read_sw(reg)		readl(((volatile void *)RTL_SW_CORE_BASE) + reg)
```
any reason your not using `ioread32`, which is preferred for new code?

plappermaul replied:
and how does it work with volatile?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	}
+};
+
+#define REG_SET(_rate, _ctrl0, _ctrl1)		\
```
i'm personally not a fan of these un-namespaced defines, yes it won't leak, but adding RTL83xx infront of it won't hurt either.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+};
+
+static const struct rtcl_reg_set rtcl_838x_cpu_reg_set[] = {
+	REG_SET(300000000, 0x045c8, 0x1414530e),
```
do we have _any_ idea what any of these bitfields mean? I doubt it right? maybe add a comment that you pulled these ad-verbatim from the SDK and we do not know what these fields mean?

plappermaul replied:
of course a big idea: [look at WIKI](https://svanheule.net/realtek/maple/register/pll_cpu_ctrl0)

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+};
+
+static const struct rtcl_reg_set rtcl_839x_mem_reg_set[] = {
+	REG_SET(100000000, 0x041CC, 0x00000000),
```
stay consistent with your capitlization of your hexes; lower-case is generally preferred)

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+extern int	rtcl_839x_dram_size;
+
+extern void	(*rtcl_839x_dram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
+static void	(*rtcl_839x_sram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
```
maybe put these in a header with the same name as your sram.S? makes it clear what and where they are from.

plappermaul replied:
Added a comment but will not change that.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	unsigned int ctrl1 = read_sw(rtcl_regs[rtcl_ccu->soc][REG_CTRL1][clk->idx]);
+
+	unsigned int div1, div2 = rtcl_xdiv[clk->idx];
+	unsigned int cmu_sel_prediv = 1 << RTL_PLL_CTRL0_CMU_SEL_PREDIV(ctrl0);
```
```suggestion
	unsigned int cmu_sel_prediv = BIT(RTL_PLL_CTRL0_CMU_SEL_PREDIV(ctrl0));
```

plappermaul replied:
leave as is.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	case SOC_RTL839X:
+		cmu_divn2_selb = RTL839X_PLL_CTRL1_CMU_DIVN2_SELB(ctrl1);
+		cmu_divn3_sel = rtcl_divn3[RTL839X_PLL_CTRL1_CMU_DIVN3_SEL(ctrl1)];
+		break;
```
```suggestion
		break;
	default:
		dev_err(hw->dev, "Unable to set rate '%d' for unknown SoC ID '%d'\n", rate, rtcl_ccu->soc);
		return -EINVAL;
```

plappermaul replied:
This will not happen as rtcl_ccu_create() will fail during initialization.

oliv3r replied:
But that's making assumptions; stuff may get refactored, and this function may be called with different understanding context. Being defensive here, adds 1/2 lines of code, pleases some linters (all cases caught) and potentially catches a 1 in a million bug.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	}
+	div1 = cmu_divn2_selb ? cmu_divn3_sel : cmu_divn2;
+
+	return (((parent_rate / 16) * cmu_ncode_in) / (div1 * div2)) *
```
can we define the '16' to make it less magical? what does 16 stand for? some fixed prediv?

plappermaul replied:
parent_rate (25000000) * cmu_ncode_in (4-260) might be > 2^32. This is to avoid 64 bit arithmetic. See * 16 compensation on next line.

oliv3r replied:
smart thing to do; wonder if there's some existing macro's for this like @svanheule pointed out earlier? But then leave this as a comment to make it easier for future you :p

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	spin_lock_irqsave(&rtcl_ccu->lock, irqflags);
+	rtcl_838x_sram_set_rate(clk_idx, reg->ctrl0, reg->ctrl1);
+	spin_unlock_irqrestore(&rtcl_ccu->lock, irqflags);
+	return 0;
```
```suggestion

	return 0;
```

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	unsigned long irqflags;
+
+	spin_lock_irqsave(&rtcl_ccu->lock, irqflags);
+	rtcl_838x_sram_set_rate(clk_idx, reg->ctrl0, reg->ctrl1);
```
from the name of this function, I would be included to think you are setting the rate of the SRAM itself. I assume that is not true, you are setting the rate using the sram-blob? setting a variable within the sram to later execute? I don't have a better name at this moment either however :(

plappermaul replied:
Functions commented as above. but will keep the names.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+
+	spin_lock_irqsave(&rtcl_ccu->lock, irqflags);
+	mtflags = dmt();
+	vpflags = dvpe();
```
do we have more verbose names for these functions? nobody will remember what these functions do based on the name :(

plappermaul replied:
Nope. Added a small comment.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+static int rtcl_839x_set_rate(int clk_idx, const struct rtcl_reg_set *reg)
+{
+	unsigned long irqflags;
+	unsigned int mtflags, vpflags;
```
unless these are very tightly coupled; one definition per variable is preferred.

any reason we don't want to use any of the fixed-size uints?

plappermaul replied:
type comes from function definition and all these are needed to preserve states so they are VERY tightly coupled.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	}
+
+	tab_idx = (rate - rtcl_round_set[rtcl_ccu->soc][clk->idx].min)
+		/ rtcl_round_set[rtcl_ccu->soc][clk->idx].step;
```
if your line is not MUCH longer then 80 characters, don't be affraid to keep it on a single line.

in this case, I'd personally probably od it more like:
```suggestion
	tab_idx = rate;
	tab_idx -=rtcl_round_set[rtcl_ccu->soc][clk->idx].min;
	tab_idx /= rtcl_round_set[rtcl_ccu->soc][clk->idx].step;
```
which SLIGHTLY helps in readability (as it un-nests it somewhat).

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	unsigned long rrate = max(clk->min, min(clk->max, rate));
+	const struct rtcl_round_set *round = &rtcl_round_set[rtcl_ccu->soc][clk->idx];
+
+	rrate = ((rrate + (round->step >> 2)) / round->step) * round->step;
```
is this a 2 because of a '/ 4`? then let the compiler do this optimization. As this line is all about math, keep math as math, makes it easier to read.

plappermaul replied:
fixed to a ">> 1". we want to round up/down here.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+ */
+
+#define COMPAT_RTL838X "realtek,rtl8380-clock"
+#define COMPAT_RTL839X "realtek,rtl8390-clock"
```
is this a thing now? :p probably not do this, so that 'grep' works better

plappermaul replied:
In my opinion using a constant 4 times thorughout the code justifies a define. But if you like.

oliv3r replied:
I think the whole idea is that the string IS the constant is kinda the idea behind compatible strings. I get what you're saying though; its just not commonly done, probably for `git grep` reasons.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	int soc;
+
+	if (of_device_is_compatible(np, COMPAT_RTL838X))
+		soc = SOC_RTL838X;
```
while this is perfectly fine; it is also often done that you assign your data structure in this case, your data structure might only contain this single entry now, but 'that's okay', it allows for easier and cleaner expantion later, making adding things that are specific to a device (min/max rates for example) a lot easier.

plappermaul replied:
As it is perfectly fine I will keep it for clean code.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+		return -ENOMEM;
+
+	rtcl_ccu->soc = soc;
+	rtcl_ccu->dram.type = RTL_MC_MCR_DRAMTYPE(read_soc(RTL_MC_MCR));
```
so this is a bit weird, because normally as a platform driver, you could call one of the of_ functions, that will give you the address you need, as per devicetree definition. I haven't seen where you call rtl_ccu_create, and this might be a mips/clock thing, but usually 'this is not the way'. I think ralink was also mips wasn't it? While they have some nice regmap driver, they also get the address in a sense from the dt (https://elixir.bootlin.com/linux/latest/source/drivers/clk/ralink/clk-mt7621.c#L492) (i lack proper examples that are not arm :p)

plappermaul replied:
I will keep my patch simple and will not fix missing memory controller DT registers.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	int ret;
+	const struct rtcl_round_set *round = &rtcl_round_set[rtcl_ccu->soc][clk_idx];
+	struct rtcl_clk *rclk = &rtcl_ccu->clks[clk_idx];
+	struct clk *clk = clk_get(NULL, rtcl_clk_info[clk_idx].name);
```
i think before, you where consistent with your variable deifnitions. I like to keep 'some' order. Be it alphabetical, length, etc. This makes it _a little_ more easy to spot duplicates or problems.

plappermaul replied:
will stick to length.

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+/*
+ * Determine startup clock rates by default framework. For min/max values we will mix the DT
+ * settings with the internal tables. DDR3 design allows a minimum of 303 MHz. To get below that
+ * barrier we must switch the memory controller into a DLL off/disable mode. Keep the complexity
```
DLL or PLL? What is a DLL?

svanheule replied:
https://en.wikipedia.org/wiki/Delay-locked_loop

oliv3r replied:
So lets write it out for clarity; as there's plenty of [abbreviations](https://www.abbreviations.com/DLL) even in programming :)

plappermaul replied:
no longer there ...

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	rclk->min = max_t(unsigned long, vals[0], round->min);
+	rclk->max = min_t(unsigned long, vals[1], round->max);
+
+	if (clk_idx == CLK_MEM && rtcl_ccu->dram.type == MEM_DDR3) {
```
```suggestion
	if ((clk_idx == CLK_MEM) && (rtcl_ccu->dram.type == MEM_DDR3)) {
```

plappermaul replied:
no longer there ...

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	hw_init.name = rtcl_clk_info[clk_idx].name;
+	hw_init.ops = &rtcl_clk_ops;
+	hw_init.parent_data = &parent_data;
+	hw_init.num_parents = 1;
```
we don't have some 'size' or 'count' from our parents do we? clk_idx is probably also not useful is it?

plappermaul replied:
?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	unsigned int idx = clkspec->args[0];
+
+	if (idx >= CLK_COUNT) {
+		pr_err("%s: Invalid index %u\n", __func__, idx);
```
can we get a 'dev' from clkspec or prv?

plappermaul replied:
All three pr_err() are in early initialization where we totally miss a device. Like all other combined clock/platform drivers.

oliv3r replied:
We can't get a de from  `of_phandle_args->np` can we?

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+	ret = of_clk_add_hw_provider(rtcl_ccu->np, rtcl_get_clkhw, rtcl_ccu);
+	if (!ret)
+		return 0;
+
```
does it make sense to swap the two? if (ret) -> print error, goto error; and then 'return 0; for success? almost reads like we return 0 on error ... reads a little bit as an anti-pattern.

plappermaul replied:
If you like

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+{
+	int ret;
+
+	if (!rtcl_ccu) {
```
not sure how happy is with these kind of globals, but I understand you are in a bit of an iffy situation with the legacy setup ...

plappermaul replied:
so we are :(

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -0,0 +1,750 @@
+		.name = "rtl83xx-clk",
+		.of_match_table = rtcl_dt_ids,
+	},
+	.probe = rtcl_probe_late,
```
if this is the regular platform probe function, why is it called 'late'?

plappermaul replied:
because of the probe_early() functions lines above

oliv3r commented on target/linux/realtek/patches-5.10/318-add-rtl83xx-clk-support.patch:
```diff
@@ -0,0 +1,26 @@
+  obj-$(CONFIG_MACH_PISTACHIO)		+= pistachio/
+  obj-$(CONFIG_COMMON_CLK_PXA)		+= pxa/
+  obj-$(CONFIG_COMMON_CLK_QCOM)		+= qcom/
++ obj-y					+= realtek/
```
shouldn't this read
```suggestion
+ obj-$(CONFIG_COMMON_CLK_REALTEK)					+= realtek/
```
You now compile this always (which is 'okay' for our platform of course, so if that's the reason, that it makes it much simpler/easier  ... unless of course we have defines under realtek that prevent this from happening of course, but doesn't seem to be the common case ...

oliv3r commented on target/linux/realtek/rtl839x/config-5.10:
```diff
@@ -12,7 +12,8 @@ CONFIG_CLKDEV_LOOKUP=y
 CONFIG_CLONE_BACKWARDS=y
 CONFIG_COMMON_CLK=y
-CONFIG_COMMON_CLK_BOSTON=y
+# CONFIG_COMMON_CLK_BOSTON is not set
```
probably very valid; probably should be in its own cosmetic comment :(

plappermaul replied:
dropped

oliv3r commented on target/linux/realtek/files-5.10/drivers/clk/realtek/Kconfig:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: GPL-2.0-only
+config COMMON_CLK_RTL83XX
+	bool "Clock driver for Realtek RTL83XX"
+	depends on RTL83XX
```
```suggestion
	depends on COMMON_CLK_REALTEK && RTL83XX
```

and then one above 'enable support for realtek clocks to have this a sub-menu

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/setup.c:
```diff
@@ -66,10 +64,41 @@ void __init plat_time_init(void)
+
+	timer_probe();
+}
+
```
this new version is a lot simpler and a lot less legacy; chapeau!

oliv3r commented on target/linux/realtek/dts-5.10/rtl838x.dtsi:
```diff
@@ -60,25 +62,43 @@
-		#clock-cells = <0>;
-		clock-frequency = <200000000>;
+	chosen {
+		bootargs = "console=ttyS0,115200";
```
this shouldn't be here should it? For one, 'chosen' should be much higher (nodes are alphabetically sorted) but you might want to keep this in a generic commit

plappermaul replied:
this diff is STRANGE as I did not touch chosen. See a few linesa above where it was deleted.

oliv3r replied:
probably git being dumb. Just double check its sorted in the correct location; that might have triggered git to think you moved it.

plappermaul commented:
PR v7 includes:

- switch to OPP tables
- enable cpufreq-dt + userspace governor

To sky-rocket over the 200 comments I once again exspect tons of nit-picky feedback. Nevertheless it would be helpful if someone can really test this bunch of code.

Borromini commented:
To be on the safe side: is wgetting https://patch-diff.githubusercontent.com/raw/openwrt/openwrt/pull/10351.patch the right way to get the v7? I've noticed with other PRs that that approach might not always yield the most recent revision. Thanks.

Other than that: anything specific that should be tested or verified?

plappermaul commented:
Looks good. URL is most recent version.

Regarding the test: Do whatever you like and how much time you can spare.

Borromini commented:
So... I tried playing with the scaling governor :smile:. I haven't connected serial yet, so I can't see what's happening. Do see it locking up though - and then it reboots.

```
root@OpenWrt:~# cat /sys/bus/cpu/devices/cpu0/cpufreq/scaling_governor 
userspace
root@OpenWrt:~# cat /sys/devices/system/cpu/cpufreq/policy0/scaling_available_governors 
ondemand userspace powersave performance 
root@OpenWrt:~# echo 'ondemand' > /sys/devices/system/cpu/cpufreq/policy0/scaling_governor
```

Am I approaching this the wrong way or is it not supposed to work yet?

plappermaul commented:
Thanks for testing and sorry for the inconvenience. Two questions:

- What SOC are you testing on 838x oder 839x?
- Cann you set frequency manual? echo 450000 > scaling_setspeed

Borromini commented:
> Thanks for testing and sorry for the inconvenience. 

Hey, no worries. That's what testing is for right.

>Two questions:
>     * What SOC are you testing on 838x oder 839x and what model?
>     * Cann you set frequency manual? echo 450000 > scaling_setspeed

RTL838x (ZyXEL GS1900-8HP v1).

Setting it manually seems to work:
```
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# cat scaling_governor
userspace
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# cat scaling_cur_freq
500000
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# echo 450000 > scaling_setspeed
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# cat scaling_cur_freq
450000
```

plappermaul commented:
I took a look at the ondemand governor. From my observation it switches clocks frequently and and as we need busy waiting loops we consume some CPU during switching. That results in continous system CPU usage (20%) on my RTL839X. I will dig into this.

Until then you could check two things:

- CPU performance on lowest vs. highest frequency.
- Wattage difference between lowest and highest frequency

Thanks in advance.

Neustradamus commented:
To follow this PR :)

Borromini commented:
> Until then you could check two things:
>   * CPU performance on lowest vs. highest frequency.

Any suggestions on how to do this? Iperf between clients with the switch in between?

>   * Wattage difference between lowest and highest frequency

I have a cheap energy meter I can put in between, but it's unreliable in the lower range (10W or less) AFAIK. Short of professional equipment, I suppose there's no better way to measure this?

plappermaul commented:
v8 enhancements

- improved 838x stability
- improved 839x stability
- include only required asm sources into build

plappermaul commented:
Finally I was able to run a very simple benchmark on a 839x. Copy /dev/zero via SCP to external PC

```
Frequency   Grade       Transfer    Config
425 MHz     Minimum     1.26 MB/s   echo 425000 > scaling_setspeed
700 MHz     Default     1.85 MB/s   echo 700000 > scaling_setspeed
750 MHz     Overclock   1.95 MB/s   echo 750000 > scaling_setspeed
800 MHz     Overclock   2.04 MB/s   echo 800000 > scaling_setspeed
```

svanheule commented:
Merged the platform init patch ahead of the other changes, since that's useful to have in any case.

396e190f0be7 realtek: more generic platform initialization

svanheule commented:
On RTL8381M (GS110TPP v1) my results are a bit strange (gzip a 10MB file of random data):
```
@ 500 MHz
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# time cat /tmp/null.bin | gzip > /dev/null
real	0m 20.90s
user	0m 0.01s
sys	0m 0.17s
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# cat cpuinfo_cur_freq
500000

@ 400 MHz
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# echo 400000 > scaling_setspeed 
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# time cat /tmp/null.bin | gzip > /dev/null
real	0m 18.84s
user	0m 0.00s
sys	0m 0.12s

@ 350 MHz
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# echo 350000 > scaling_setspeed 
root@OpenWrt:/sys/devices/system/cpu/cpufreq/policy0# time cat /tmp/null.bin | gzip > /dev/null
real	0m 17.84s
user	0m 0.00s
sys	0m 0.18s
```

Not sure how to interpret this. Is wall time incorrect once we reclock the CPU? Is the relative time spent on overhead going down because the CPU's cycle take longer, causing the operation to virtually speed up?

When changing the CPU frequency, I also got the following notification on the first change:
```
[  198.500000] R4K timer is unstable due to CPU frequency change
```

All SoC, from RTL838x to RTL931x, have external timers that are driven by the Lexra bus clock, so we should be able to use those if we want accurate wall times.

plappermaul committed:
realtek: add PLL DT binding includes

Add some constants for sharing between DT and drivers.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul commented:
PR v9 contains a slightly modified final version of the clock driver. It will provide reading of all clocks on RTL838X and RTL839X but only allows changing the CPU clock on both device familys. MEM changes would require additional DTR/MCR/DLL programming and that is beyond of spare time for this project. LXB changing has only been tested rudimentary and has currently no use case.

plappermaul committed:
realtek: add RTL83XX clock driver

Add a new self-contained combined clock & platform driver that allows to access the PLL
hardware clocks of RTL83XX devices. Currently it provides info about CPU, MEM and LXB
clocks on RTL838X and RTL839X devices and additionally allows to change the CPU clocks.
Changing the clocks multiple times on a DGS-1210-20 and a DGS-1210-52 already works well
and is multithreading safe on the RTL839X. Even a cpufreq initiated change of the CPU
clock works fine. Loading the driver will add some meaningful logging.

[0.000000] rtl83xx-clk: initialized, CPU 500 MHz, MEM 300 MHz (8 Bit DDR3), LXB 200 MHz
[0.279456] rtl83xx-clk soc:clock-controller: rate setting enabled, CPU 325-600 MHz,
           MEM 300-300 MHz, LXB 200-200 MHz, OVERCLOCK AT OWN RISK

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: add patch to enable new clock driver in kernel

Allow building the clock driver with kernel config options.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: enable basic config for cpufreq framework

A new clock driver makes more sense if it can be used from consumers like cpufreq.
Before we enable the driver we must tell the config that the RTL838X and RTL839X
targets allow CPU frequency changing.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: activate clock driver for RTL838X/RTL839X targets

Make use the new clock driver for RTL838X and RTL839x target devices. Of course
we will enable their primary consumer (cpufreq-dt) too. To be careful just set
the default governor to userspace. As we rely on SRAM activate that module too.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: switch RTL838X/RTL839X DT to new clock driver

Use new DT clockdriver syntax for RTL838X/RTL839X targets. To make it work
we need to change some nodes:
- define the external oscillator speed (25MHz)
- define SRAM
- add clock controller
- Add second CPU for RTL839X
- map all devices to new clocks
- Remove dummy LXB clock
- add CPU OPP table

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

svanheule commented:
Merged with the following commits:
1efaad03bb20 realtek: add PLL DT binding includes
4850bd887c3a realtek: add RTL83XX clock driver
800d5fb3c6a1 realtek: add patch to enable new clock driver in kernel
5df36d484968 realtek: enable basic config for cpufreq framework
7c18aab6e051 realtek: activate clock driver for RTL838X/RTL839X targets
48f3746fe5a9 realtek: switch RTL838X/RTL839X DT to new clock driver

checkpatch.pl was still complaining about 4850bd887c3a ("realtek: add RTL83XX clock driver"), so I cleaned that up for you. I've also updated the commit message for commit 5df36d484968 ("realtek: enable basic config for cpufreq framework"), to explain the caveats of allowing CPU reclocking at this moment.
